### PR TITLE
CLI: Fixes #1728 mv command works on notebooks now

### LINKED ISSUE #1728: Create subnotebooks in terminal
### Operating system

- macOS
- Linux

### Application

- Terminal

Just wondering if there was a way to create new subnotebooks in the terminal app? Was looking through the [terminal docs](https://joplinapp.org/terminal/) and was not able to find that feature.

### PR DESCRIPTION
**Pull request by kai11**

### CODE REVIEW

**Comment by tessus** [comment]
I am a bit puzzled by how you get the ids in the terminal app. You can't see them afaik.

Also, how to you handle ambiguous (names) source or target note/notebook?

**Comment by laurent22** [comment]
As Tessus mentioned we need to disambiguate when there are multiple notebooks or notes with the same title. What can happens is way to unpredictable otherwise because they query will search all notes in all notebooks and sub-notebooks and pick the first one it finds. There's a `loadItem()` function in app.ts that simply throws an exception when there's an ambiguity, which perhaps is good enough, as it just means the user needs to enter the ID instead.

File: packages/app-cli/app/command-mv.js [hunk 1]
```diff
@@ -7,25 +7,38 @@ const Note = require('@joplin/lib/models/Note').default;
...
-		if (!folder) throw new Error(_('Cannot find "%s".', destination));
-
-		const notes = await app().loadItems(BaseModel.TYPE_NOTE, pattern);
-		if (!notes.length) throw new Error(_('Cannot find "%s".', pattern));
+		// TODO: allow partial ids
+		let folder = await Folder.loadByField('id', destination);
+		if (!folder) {
+			folder = await Folder.loadByField('title', destination);
+			if (!folder) throw new Error(_('Cannot find "%s".', destination));
+		}
```
Comment (laurent22) [inline packages/app-cli/app/command-mv.js:26]: Aren't partial IDs already handled by loadItem/loadItems in app.js? The problem is that you use Folder.loadByField and thus bypass all the business logic implemented in the app().loadItems call you removed.

**Comment by tessus** [comment]
>  the user needs to enter the ID instead.

How? There's no way to get the ids for folders. I think the ids for notes you can get by checking the metadata.

I don't understand why this dude never created a PR.... https://github.com/DominicLavery/joplin/commit/6428e669b294c6225d8ec487bc00bd0597765a23

**Comment by kai11** [comment]
# Testing plan
- create following structure
  - folder1
    - note1
    - note2
    - note2
  - folder2
  - folder3
    - note3
  - folder3
```
mkbook folder1
mknote note1
mknote note2
mknote note2
mkbook folder2
mkbook folder3
mknote note3
mkbook folder3
```

- select folder1
- `mv note1 folder0` must fail (target don't exist)
- `mv note0 folder2` must fail (source note don't exist)
- `mv folder0 folder2` must fail (source folder don't exist)
- `mv note1 folder3` - must fail (moving to ambiguous folder)
- `mv folder3 folder1` must fail (moving ambiguous source folders)
- `mv note3 folder1` - must fail (moving note not from currently selected folder)
- `mv note1 folder2` (moving one note using title)
- `mv folder1 folder2` (moving one folder using title)
- `mv note2 folder2` (moving ambiguous note must be successful - but ONLY ONE note is moved. This is current behavior but is it correct?)
- `mv folder1 root` (make folder1 to root)
- select folder2
- `mv note* folder1` (moving using pattern)
- run shortcut 'toggle metadata' - tm by default
- select folder1\note1, write down note id (nid1) and parent id (pid1)
- select folder3\note3, write down note id (nid3) and parent id (pid3)
- select folder3
- `mv partial_nid1 partial_pid3` (move note using partial id, however note is "Untitled" until I change folder back and fourth).
- `mv partial_pid1 partial_pid3` (move folder using partial id to folder id)

Problems \ Questions:
- moving ambiguous notes is successful - but ONLY ONE note is moved. This is current behavior but is it correct?
- When moving note to currently selected folder (only by note id, by title it's impossible since this search performed in current folder only), this note become "Untitled" until I change folder back and fourth. I don't know how to fix this.

**Comment by tessus** [comment]
> moving ambiguous notes is successful - but ONLY ONE note is moved. This is current behavior but is it correct?

That's a problem, because which one is picked? We need to find a solution for that. A warning that one has to use the id instead for example. The same is true for ambiguous folders.

**Comment by kai11** [comment]
> The same is true for ambiguous folders.
Already done. Twice.
https://github.com/kai11/joplin/blob/dev/packages/app-cli/app/command-mv.js#L27
https://github.com/kai11/joplin/blob/dev/packages/app-cli/app/command-mv.js#L34
Is it correct method?
I did not found any API call for that.

> That's a problem, because which one is picked?
I think both is better than random. This is infrequent situation anyway - two different notes with same title in same notebook is probably a user mistake.

Also note name with "*" WILL be treated like pattern and will move both as expected.

And how to fix Untitled?

**Comment by kai11** [comment]
@laurent22 @tessus I cannot proceed. See my questions above.

**Comment by tessus** [comment]
As mentioned before, if the source or target is ambiguous (because the title of the note or name of the notebook are not unique), an error should be presented to the user that the id has to be used instead. No other action can be taken. You can't move anything in that case, because which one is picked? It's like rolling a dice.

This https://github.com/laurent22/joplin/issues/1728#issuecomment-687479331 has a suggestion how to maybe find out the id.

**Comment by kai11** [comment]
@tessus Your answer is not helpful. I'm asking about Joplin internal API, not UI and not UX.

Im NOT going to show IDs like Laurent suggested, since you can always use $n and it will be replaced with id of current note and its unambiguous. And its not my goal for this PR.

Also I changed my mind on ambiguous notes - I'll show error like I do for folders.

My questions:
1. Is it correct call to check for duplicated folders?  
https://github.com/kai11/joplin/blob/dev/packages/app-cli/app/command-mv.js#L27
https://github.com/kai11/joplin/blob/dev/packages/app-cli/app/command-mv.js#L34
I did not found any API call for that.
With this call all folder-related entries in my test plan above pass successfully.

2. Same question - how to check for dups - for Notes. I'm going to use Note.search to check for duplicates. Is it the best choice? 

3. Also, for notes I need to avoid dup check if I see "*" in search string. Is it correct? Is it the only pattern Joplin supports?

4. Test case from test plan above:
```mv partial_nid1 partial_pid3``` (move note using partial id, however note is "Untitled" until I change folder back and fourth).
I dont know how to fix this Untitled problem. How I can refresh list of notes?

CC @laurent22

**Comment by tessus** [comment]
I think we are talking past each other.

**Comment by kai11** [comment]
I won't finish this PR. 
Few notes for others who decide to pull this through or patch their own Joplin cli:
- If I recall correctly, listItem return ONE folder and dont check for duplicates.
- listItems have convoluted logic and don't return multiple folders
- My test plan above covers all test cases tessus wanted and was designed to run sequentially on empty database. Subsequent tests expect previous tests (successful ones of course) to be completed successfully.
- diff below.


```
diff --git a/packages/app-cli/app/command-mv.js b/packages/app-cli/app/command-mv.js
index 4879e38f8..a2ca50bf6 100644
--- a/packages/app-cli/app/command-mv.js
+++ b/packages/app-cli/app/command-mv.js
@@ -7,25 +7,45 @@ const Note = require('@joplin/lib/models/Note').default;
 
 class Command extends BaseCommand {
 	usage() {
-		return 'mv <note> [notebook]';
+		return 'mv <item> [notebook]';
 	}
 
 	description() {
-		return _('Moves the notes matching <note> to [notebook].');
+		return _('Moves the given <item> (notes matching pattern in current notebook or one notebook) to [notebook]. If <item> is subnotebook and [notebook] is "root", will make <item> parent notebook');
 	}
 
 	async action(args) {
-		const pattern = args['note'];
+		const pattern = args['item'];
 		const destination = args['notebook'];
+		let folder = null;
 
-		const folder = await Folder.loadByField('title', destination);
-		if (!folder) throw new Error(_('Cannot find "%s".', destination));
+		if (destination != 'root') {
+			folder = await app().loadItem(BaseModel.TYPE_FOLDER, destination);
+			if (!folder) throw new Error(_('Cannot find "%s".', destination));
+		}
 
-		const notes = await app().loadItems(BaseModel.TYPE_NOTE, pattern);
-		if (!notes.length) throw new Error(_('Cannot find "%s".', pattern));
+		const dstDups = await Folder.search({ titlePattern: destination, limit: 2 });
+		if (dstDups.length > 1) {
+			throw new Error(_('Ambiguous notebook "%s". Please use notebook id instead - look for parent id in metadata', destination));
+		}
 
-		for (let i = 0; i < notes.length; i++) {
-			await Note.moveToFolder(notes[i].id, folder.id);
+		const itemFolder = await app().loadItem(BaseModel.TYPE_FOLDER, pattern);
+		if (itemFolder) {
+			const srcDups = await Folder.search({ titlePattern: pattern, limit: 2 });
+			if (srcDups.length > 1) {
+				throw new Error(_('Ambiguous notebook "%s". Please use notebook id instead - look for parent id in metadata or use $b for current', pattern));
+			}
+			if (destination == 'root') {
+				await Folder.moveToFolder(itemFolder.id, '');
+			} else {
+				await Folder.moveToFolder(itemFolder.id, folder.id);
+			}
+		} else {
+			const notes = await app().loadItems(BaseModel.TYPE_NOTE, pattern);
+			if (!notes.length) throw new Error(_('Cannot find "%s".', pattern));
+			for (let i = 0; i < notes.length; i++) {
+				await Note.moveToFolder(notes[i].id, folder.id);
+			}
 		}
 	}
 }
```

**Pull request closed without merging**